#### Дмитрий, привет! 

Меня зовут Максим Попов, и я буду проверять твой проект. Давай будем общаться на «ты». Но если тебе комфортнее на «вы», обязательно сообщи мне об этом. 
По ходу тетрадки ты найдешь комментарии к проекту, они будут в отдельных ячейках Markdown с заголовком «Комментарий ревьюера». Постарайся учесть эти комментарии для дальнейших проектов.

---

Я буду отмечать комментарии цветом и примечанием. В конце заголовка комментария будет указан номер итерации проверки. Пожалуйста, не удаляй и не правь мои комментарии

Критично - требует исправления, влияет на удачное выполнение проекта

Рекомендация - комментарий является рекомендацией или советом. Mожешь учесть их при выполнении будущих заданий. На твое усмотрение

Отлично - Так отмечены удачные решения и подходы, на которые стоит опираться в будущих проектах


---
<div class="alert alert-block alert-danger">
    

### Комментарий от ревьюера
Критично

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">
    

### Комментарий от ревьюера
Рекомендация

Такой комментарий является рекомендацией или советом. Mожешь учесть их при выполнении будущих заданий. На твое усмотрение. 
</div>

---

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера
Отлично
    
Так отмечены удачные решения и подходы, на которые стоит опираться в будущих проектах
</div>


Желательно реагировать на комментарии ('исправлено', 'не понятно как исправить ошибку' и т.п.) - так мы сможем более эффективно поработать над этим проектом.

---

Будет здорово, если ты будешь помечать свои действия следующим образом:



<div class="alert alert-block alert-info">

### Комментарий студента:

</div>

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера v_1
    
Дмитрий, ты отлично справляешься с запросами, ошибка, которую нашел, неочевидна.
    
Обращу твое внимание на следующие моменты:
    
 - У тебя достаточно чистый код, но не соблюдаешь отступы, рекомендую ознакомиться с общепринятым стилем
    
https://www.sqlstyle.guide/ru/

- Здорово, что заворачиваешь подзапросы в CTE. Так намного проще и читать, и писать код.
    
    
Если возникнут вопросы ко мне, можешь оставить их в отдельном комментарии. А также пиши, если есть желание детальнее познакомиться с SQL, поделюсь полезными материалами.
        
</div>

<div class="alert alert-block alert-info">

### Комментарий студента: привет!Спасибо за ссылку, обязательно посмотрю. С sql у меня точно есть пробелы и после основного обучения я бы хотел заняться им плотнее, поэтому буду рад любым ссылка и метериалам.

</div>

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_2
    
Дмитрий, привет!
    
Отлично, все критичные комментарии отработаны!
    
У тебя получился отличный проект, успеха в поиске работы!
    
Для дальнейшего изучения рекомендую следующие ресурсы:
    
1. https://datalearn.ru/kurs-po-sql 
    
Сразу переходи к оконкам и MS SQL Server, там много полезной инфы будет уже про создание таблиц, хранимые процедуры и т. п. 
    
2. https://stepik.org/course/3203/promo (это уже если хочется изучить поглубже)
    
Тут уже про то, что под капотом SQL (было бы неплохо ориентироваться в алгоритмах для прохождения, очень рекомендую книгу "Грокаем алгоритмы", чтобы изучить основы, вдохновиться и понять, что алгоритмы - это не сложно, а увлекательно и красиво)
        
</div>

# Подготовка нового продукта для сервиса чтения книг по подписке
---

## Анализ данных о книгах

После покупки нашей компанией сервиса чтения книг по подписке необходимо проанализировать имеющиеся данные о книгах, издательствах, авторах, пользовательских обзорах книг, чтобы сформулировать ценностное предложение для нового продукта.

**Цель исследования:** на основе анализа данных сформировать предложения для нового продукта.

**Задачи исследования:**
1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

В наличии есть следующие данные:

**Исходные данные:**
Таблица `books` содержит данные о книгах.

**Поля таблицы:**
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors` содержит данные об авторах.

**Поля таблицы:**
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers` содержит данные об издательствах.

**Поля таблицы:**
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

Таблица `ratings` содержит данные о пользовательских оценках книг.

**Поля таблицы:**
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews` содержит данные о пользовательских обзорах на книги.

**Поля таблицы:**
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

Таблицы взаимосвязаны следующим образом:
![image.png](attachment:image.png)

# Содержание
0. [Загрузка необходимых библиотек](#libs)
1. [Подключение к базе данных](#connect)
2. [Знакомство с данными](#view)
3. [Расчеты](#count)
  - 3.1 [Определение кол-ва книг, вышедших после 1 января 2000 года](#count1)
  - 3.2 [Определение кол-ва обзоров и средней оценки каждой книги](#count2)
  - 3.3 [Издательство с наибольшим кол-вом выпущенных больших книг](#count3)
  - 3.4 [Определение автора с самой высокой средней оценкой](#count4)
  - 3.5 [Расчет среднего кол-ва обзоров по пользователям](#count5)
4. [Выводы](#result)

# 0. Загрузка необходимых библиотек <a name="libs"></a>

Загрузим библиотеки, которые будем использовать в исследовании.

In [3]:
#загрузка библиотек
import pandas as pd
from IPython.display import display
from sqlalchemy import create_engine 

# 1. Подключение к базе данных <a name="connect"></a>

Настроим подключение к БД.

In [4]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

# 2. Знакомство с данными <a name="view"></a>

Загрузим данные из БД и посмотрим на их структуру.

In [5]:
#сделаем функцию для загрузки данных
def data_load(table_name):
    '''Функция получает название таблицы, из которой берутся данные, и загружает эти данные в датафрейм'''
    query = '''SELECT * FROM ''' + table_name 
                
    
    data =  pd.io.sql.read_sql(query, con = engine)
    
    return data

In [6]:
#загрузим данные в таблицы
books = data_load('books')
authors = data_load('authors')
publishers = data_load('publishers')
ratings = data_load('ratings')
reviews = data_load('reviews')

In [7]:
#выведем первые строки таблиц
display(books.head(10))
display(authors.head(10))
display(publishers.head(10))
display(ratings.head(10))
display(reviews.head(10))

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера
Отлично
    
Доступ к базе получили, на таблицы посмотрели, переходим к основной работе
    
</div>

# 3. Расчеты <a name="count"></a>

## 3.1 Определение кол-ва книг, вышедших после 1 января 2000 года <a name="count1"></a>

Определим, сколько книг вышло после 1 января 2000 года.

In [8]:
#пишем запрос
query = '''SELECT COUNT(distinct b.book_id)
        FROM books b
        where b.publication_date > cast('2000-01-01' as date)'''

In [9]:
#выполняем его
pd.io.sql.read_sql(query, con = engine)

count
0    819

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

    
👍🏻
</div>

**Вывод**

В нашем датасете с книгами есть **819** книг, которые вышли после 1 января 2000 года.

## 3.2 Определение кол-ва обзоров и средней оценки каждой книги <a name="count2"></a>

Для каждой книги рассчитаем количество обзоров и среднюю оценку.

In [16]:
#пишем запрос
query = '''
        select  b.book_id
                ,b.title
                ,count(rw.review_id) as count_reviews --расчет кол-ва обзоров на книгу
                ,avg(rt.rating) as avg_rate --расчет среднего рейтинга книги
        from books b
            left join ratings rt
                on b.book_id = rt.book_id
            left join reviews rw
                on b.book_id = rw.book_id
        group by b.book_id,b.title
'''

In [17]:
#выполняем его
books_rate_reviews = pd.io.sql.read_sql(query, con = engine)

In [19]:
#из любопытства выведем TOP-10 книг с наибольшим кол-вом обзоров
display(books_rate_reviews.sort_values(by = 'count_reviews', ascending = False).head(10))

book_id                                              title  \
161      948                            Twilight (Twilight  #1)   
102      750                The Hobbit  or There and Back Again   
672      673                             The Catcher in the Rye   
651      302  Harry Potter and the Prisoner of Azkaban (Harr...   
747      299  Harry Potter and the Chamber of Secrets (Harry...   
81        75               Angels & Demons (Robert Langdon  #1)   
837      301  Harry Potter and the Order of the Phoenix (Har...   
587      779  The Lightning Thief (Percy Jackson and the Oly...   
334      722  The Fellowship of the Ring (The Lord of the Ri...   
909       79                                        Animal Farm   

     count_reviews  avg_rate  
161           1120  3.662500  
102            528  4.125000  
672            516  3.825581  
651            492  4.414634  
747            480  4.287500  
81             420  3.678571  
837            375  4.186667  
587            372  4.080645  
334            370  4.391892  
909            370  3.729730

In [20]:
#из любопытства выведем TOP-10 книг с самой высокой средней оценкой среди книг, на которые написали более 50 обзоров
display(books_rate_reviews.query('count_reviews >= 50').sort_values(by = 'avg_rate', ascending = False).head(10))

book_id                                              title  \
823      581                     Snow Flower and the Secret Fan   
182       74                           And Then There Were None   
689      284                                 Gone with the Wind   
723      422                                            Matilda   
651      302  Harry Potter and the Prisoner of Azkaban (Harr...   
334      722  The Fellowship of the Ring (The Lord of the Ri...   
786       15     A Clash of Kings  (A Song of Ice and Fire  #2)   
896      830                               The Poisonwood Bible   
747      299  Harry Potter and the Chamber of Secrets (Harry...   
204      862  The Sea of Monsters (Percy Jackson and the Oly...   

     count_reviews  avg_rate  
823             55  4.636364  
182            110  4.454545  
689            175  4.428571  
723             76  4.421053  
651            492  4.414634  
334            370  4.391892  
786            115  4.391304  
896            110  4.363636  
747            480  4.287500  
204            110  4.272727

**Вывод**

Больше всего обзоров было написано на "Сумерки" - **1120**, при этом средний рейтинг у данной книги всего **3.7**. Если взять книги более чем с **50** обзорами, среди них самую высокую оценку имеет книга "Snow Flower and the Secret Fan" - **4.6**.

In [25]:
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  review_cnt  \
0      948                            Twilight (Twilight  #1)           7   
1      696             The Da Vinci Code (Robert Langdon  #2)           6   
2      207                                    Eat  Pray  Love           6   
3      627                                      The Alchemist           6   
4      673                             The Catcher in the Rye           6   
5      695  The Curious Incident of the Dog in the Night-Time           6   
6      302  Harry Potter and the Prisoner of Azkaban (Harr...           6   
7      299  Harry Potter and the Chamber of Secrets (Harry...           6   
8      497                          Outlander (Outlander  #1)           6   
9      656                                     The Book Thief           6   

   avg_rate  
0  3.662500  
1  3.830508  
2  3.395833  
3  3.789474  
4  3.825581  
5  4.081081  
6  4.414634  
7  4.287500  
8  4.125000  
9  4.264151

In [26]:
query = '''
SELECT
    books.book_id,
    books.title,
    COUNT(DISTINCT review_id) as count_reviews,
    AVG(rating) as avg_rate
FROM 
    books
INNER JOIN reviews ON  reviews.book_id = books.book_id
INNER JOIN ratings ON  ratings.book_id = books.book_id
GROUP BY
    books.book_id
ORDER BY
    review_cnt DESC
LIMIT 10
'''

In [23]:
pd.io.sql.read_sql(query, con = engine).review_cnt.sum()

61

Больше всего пользовательский обзоров (7) было написано для книги "Сумерки"

<div class="alert alert-block alert-danger">
    

    
### Комментарий от ревьюера v_1
Критично

Здесь закралась ошибка, проверь сколько у тебя всего значений в таблице с обзорами и сколько обзоров получилось в результате запроса
</div>

<div class="alert alert-block alert-info">

### Комментарий студента: Исправил, первую версию кода пока оставил

</div>

<div class="alert alert-block alert-success">

    
### Комментарий от ревьюера v_2

👍🏻
    
</div>

## 3.3 Издательство с наибольшим кол-вом выпущенных больших книг <a name="count3"></a>

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры.

In [13]:
#пишем запрос
query = ''' 
                select 
                        p.publisher
                        ,count(b.book_id) as total_books
                from books b
                    left join publishers p
                        on b.publisher_id = p.publisher_id
                where b.num_pages > 50
                group by p.publisher
                order by total_books desc
'''

In [14]:
#выполняем его
pd.io.sql.read_sql(query, con = engine).head(1)

publisher  total_books
0  Penguin Books           42

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻
    

    
</div>

**Вывод**

Наибольшее число книг толще 50 страниц выпустило издательство ***Penguin Books*** - **42** книги.

## 3.4 Определение автора с самой высокой средней оценкой <a name="count4"></a>

Определим автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками.

In [21]:
#пишем запрос
query = ''' 
        --берем книги с 50 и более оценками, считаем их средний рейтинг
        with t as (
        select  b.book_id
                ,count(rt.rating_id) as ratings
                ,avg(rt.rating) as avg_book_rate
        from books b
            left join ratings rt
                on b.book_id = rt.book_id
        group by b.book_id
        HAVING count(rt.rating_id) >= 50
        )
        
        --добавляем к предыдущей таблице авторов, рассчитываем среднюю оценку книг по автору
        select 
                a.author
                ,avg(t.avg_book_rate) as avg_author_rate  
        from t 
            left join books b
                on t.book_id = b.book_id
            left join authors a
                on b.author_id = a.author_id
        group by a.author
        order by avg_author_rate desc
'''

In [23]:
#выполняем его
pd.io.sql.read_sql(query, con = engine).head(1)

author  avg_author_rate
0  J.K. Rowling/Mary GrandPré         4.283844

**Вывод**

***J.K. Rowling/Mary GrandPré*** - автор с самой высокой средней оценкой книг (равной **4.28**) среди книг с 50 и более оценками.

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻
    

    
</div>

## 3.5 Расчет среднего кол-ва обзоров по пользователям <a name="count5"></a>

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [17]:
#пишем запрос
query = '''
        select round(AVG(sub.total_reviews)) as avg_reviews
        from 
        (select  rw.username
                ,count(rw.review_id) total_reviews
        from reviews rw
        where rw.username in (
            --отбираем пользователей, которые поставили более 50 оценок
            select rt.username
            from ratings rt
            group by rt.username
            having count(rt.rating_id) > 50
            )
        group by rw.username) as sub
'''

In [18]:
#выполняем его
pd.io.sql.read_sql(query, con = engine)

avg_reviews
0         24.0

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻

    
</div>

**Вывод**

В среднем, пользователи, которые поставили более 50 оценок, пишут **24** обзора.

# 4. Выводы <a name="result"></a>

1. **819** книг из датасета были выпущены после 1 января 2000 года;
2. Книга с наибольшим количеством обзоров - "Сумерки", среди книг с более 50 обзоров самый высокий рейтинг у "Snow Flower and the Secret Fan" - **4.6**;
3. Больше всего книг толще 50 страниц выпустило издательство ***Penguin Books*** - **42** книги;
4. ***J.K. Rowling/Mary GrandPré*** - автор с самой высокой средней оценкой книг (равной **4.28**) среди книг с 50 и более оценками;
5. В среднем, пользователи, которые поставили более 50 оценок, пишут **24** обзора.

<div class="alert alert-info"> <b>Комментарий студента1:</b>
    
Привет!

В пунктах `3.3` и `3.4` решил воспользоваться для вывода методом `head()`, чтобы не усложнять запрос без необходимости. Хотел использовать в самом запросе `top 1`, но здесь, видимо, эта функция не работает.

Спасибо за проверку!

</div>

<div class="alert alert-block alert-success">

    
### Комментарий от ревьюера v_2
    
А тебе спасибо за отличный проект!

TOP - это в MSSQL, в Postgre - LIMIT, в конце запроса))
    

    
</div>